contracts/TradeTypesLib2.sol

```diff
@@ -0,0 +1,134 @@
+/*
+ * Hubii Nahmii
+ *
+ * Compliant with the Hubii Nahmii specification v0.12.
+ *
+ * Copyright (C) 2017-2018 Hubii AS
+ */
+
+pragma solidity >=0.4.25 <0.6.0;
+
+import {MonetaryTypesLib} from "./MonetaryTypesLib.sol";
+import {NahmiiTypesLib} from "./NahmiiTypesLib.sol";
+
+/**
+ * @title     TradeTypesLib
+ * @dev       Data types centered around trade
+ */
+library TradeTypesLib2 {
+    //
+    // Enums
+    // -----------------------------------------------------------------------------------------------------------------
+    enum CurrencyRole {Intended, Conjugate}
+    enum LiquidityRole {Maker, Taker}
+    enum Intention {Buy, Sell}
+    enum TradePartyRole {Buyer, Seller}
+
+    //
+    // Structures
+    // -----------------------------------------------------------------------------------------------------------------
+    struct Operator {
+        uint256 id;
+        string data;
+    }
+
+    struct OrderPlacement {
+        Intention intention;
+
+        int256 amount;
+        NahmiiTypesLib.IntendedConjugateCurrency currencies;
+        int256 rate;
+
+        NahmiiTypesLib.CurrentPreviousInt256 residuals;
+    }
+
+    struct OrderParty {
+        uint256 nonce;
+        address wallet;
+
+        NahmiiTypesLib.IntendedConjugateCurrency balances;
+
+        string data;
+    }
+
+    struct Order {
+        OrderParty party;
+
+        OrderPlacement placement;
+
+        NahmiiTypesLib.WalletOperatorSeal seals;
+        uint256 blockNumber;
+
+        Operator operator;
+    }
+
+    struct OrderCancellation {
+        OrderParty party;
+
+        NahmiiTypesLib.WalletOperatorSeal seals;
+        uint256 blockNumber;
+
+        Operator operator;
+    }
+
+    struct TradeOrder {
+        int256 amount;
+        NahmiiTypesLib.WalletOperatorHashes hashes;
+        NahmiiTypesLib.CurrentPreviousInt256 residuals;
+    }
+
+    struct TradeParty {
+        uint256 nonce;
+        address wallet;
+
+        uint256 rollingVolume;
+
+        LiquidityRole liquidityRole;
+
+        TradeOrder order;
+
+        NahmiiTypesLib.IntendedConjugateCurrentPreviousInt256 balances;
+
+        NahmiiTypesLib.SingleFigureTotalOriginFigures fees;
+
+        string data;
+    }
+
+    struct Trade {
+        int256 amount;
+        NahmiiTypesLib.IntendedConjugateCurrency currencies;
```

Comment:
For Trades, is `IntendedConjugateCurrency` in the perspective or the buyer?

If I understand it correctly, the `intended` currency for the buyer, is the `conjugate` currency for the seller, and the seller's `intended` currency is the buyer's `conjugate`?

Reply:
The intended currency applies to both trade parties.

The _seller_ in his order wants to _sell_ intended currency _I_ for conjugate currency _C_. The _buyer_ wants at his end wants to _buy_ intended currency _I_ for conjugate currency _C_. Thus a trade is filled with buy and sell order there is agreement about the definitions of intended and conjugate currencies.